compiler/noirc_frontend/src/ownership/mod.rs

```diff
@@ -260,180 +344,321 @@ fn recur_on_parameter<'typ>(
+        } else {
+            self.handle_expression(&mut index.collection);
+        }
+
+        self.handle_expression(&mut index.index);
+
+        if clone_result && contains_array_or_str_type(&index.element_type) {
+            clone_expr(index_expr);
+        }
     }
-}
 
-fn handle_tuple(elems: &mut [Expression]) {
-    for elem in elems {
-        handle_expression(elem);
+    fn handle_cast(&mut self, cast: &mut crate::monomorphization::ast::Cast) {
+        self.handle_expression(&mut cast.lhs);
     }
-}
 
-fn handle_call(call: &mut crate::monomorphization::ast::Call) {
-    handle_expression(&mut call.func);
-    for arg in &mut call.arguments {
-        handle_expression(arg);
+    fn handle_for(&mut self, for_expr: &mut crate::monomorphization::ast::For) {
+        self.handle_expression(&mut for_expr.start_range);
+        self.handle_expression(&mut for_expr.end_range);
+        self.handle_expression(&mut for_expr.block);
     }
-}
 
-/// - Let bindings (`let _ = <expression which returns an array>;`):
-///   - Binding an array to a let binding increments the reference count of the array unless
-///     the expression is an array literal in which case it is considered to be moved.
-fn handle_let(let_expr: &mut crate::monomorphization::ast::Let) {
-    handle_expression(&mut let_expr.expression);
+    fn handle_while(&mut self, while_expr: &mut crate::monomorphization::ast::While) {
+        self.handle_expression(&mut while_expr.condition);
+        self.handle_expression(&mut while_expr.body);
+    }
 
-    if !is_array_or_str_literal(&let_expr.expression) {
-        clone_expr(&mut let_expr.expression);
+    fn handle_if(&mut self, if_expr: &mut crate::monomorphization::ast::If) {
+        self.handle_expression(&mut if_expr.condition);
+        self.handle_expression(&mut if_expr.consequence);
+        if let Some(alt) = &mut if_expr.alternative {
+            self.handle_expression(alt);
+        }
     }
-}
 
-fn handle_constrain(
-    boolean: &mut Expression,
-    msg: &mut Option<Box<(Expression, crate::hir_def::types::Type)>>,
-) {
-    handle_expression(boolean);
+    fn handle_match(&mut self, match_expr: &mut crate::monomorphization::ast::Match) {
+        for case in &mut match_expr.cases {
+            self.handle_expression(&mut case.branch);
+        }
 
-    if let Some(msg) = msg {
-        handle_expression(&mut msg.0);
+        if let Some(default_case) = &mut match_expr.default_case {
+            self.handle_expression(default_case);
+        }
     }
-}
 
-/// - Assignments (`x = <expression which returns an array>;`):
-///   - Assigning an array to an existing variable will also increment the reference
-///     count of the array unless it is an array literal.
-fn handle_assign(assign: &mut crate::monomorphization::ast::Assign) {
-    handle_lvalue(&mut assign.lvalue);
-    handle_expression(&mut assign.expression);
+    fn handle_tuple(&mut self, elems: &mut [Expression]) {
```

Comment:
```suggestion
    fn handle_tuple(&mut self, elements: &mut [Expression]) {
```
nit: or add `elems` to cspell